lib/rucio/core/transfer.py

```diff
@@ -979,115 +1028,54 @@ def get_transfer_requests_and_source_replicas(total_workers=0, worker_number=0,
-                if destination_protocol == 'https+srm':
-                    dest_url = re.sub('srm', 'https+srm', dest_url)
-
-                # III - The transfer queued previously is a multihop, but this one is direct.
+                # The transfer queued previously is a multihop, but this one is direct.
                 # Reset the sources, remove the multihop flag
                 if transfers[req_id].get('multihop', False):
                     transfers[req_id].pop('multihop', None)
                     transfers[req_id]['sources'] = []
 
-                if ranking is None:
-                    ranking = 0
-                # TAPE should not mixed with Disk and should not use as first try
-                # If there is a source whose ranking is no less than the Tape ranking, Tape will not be used.
+                current_source_is_tape = transfers[req_id]['bring_online']
+                new_source_is_tape = ctx.is_tape_rse(source_rse_id) or ctx.rse_attrs(source_rse_id).get('staging_required', False)
 
-                if rses_info[source_rse_id]['rse_type'] == RSEType.TAPE or rses_info[source_rse_id]['rse_type'] == 'TAPE' or rse_attrs[source_rse_id].get('staging_required', False):
-                    # current src_rse is Tape
-                    if not allow_tape_source:
-                        continue
-                    if not transfers[req_id]['bring_online']:
-                        # the sources already founded are disks.
-
-                        avail_top_ranking = None
-                        # avail_top_ranking stays None if there are no sources (reset if multihop)
-                        founded_sources = transfers[req_id]['sources']
-                        for founded_source in founded_sources:
-                            if avail_top_ranking is None:
-                                avail_top_ranking = founded_source[3]
-                                continue
-                            if founded_source[3] is not None and founded_source[3] > avail_top_ranking:
-                                avail_top_ranking = founded_source[3]
-
-                        if avail_top_ranking is not None and avail_top_ranking >= ranking:
-                            # current Tape source is not the highest ranking, will use disk sources
+                if new_source_is_tape and not allow_tape_source:
+                    continue
+
+                if current_source_is_tape and not new_source_is_tape or \
+                        new_source_is_tape and not current_source_is_tape:
+                    # Tape and Disk sources must not be used at the same time.
+                    # Either keep existing sources unchanged, or substitute all existing source with the new source.
+
+                    # Find the best ranking among existing sources
+                    avail_top_ranking = None
+                    for founded_source in transfers[req_id]['sources']:
```

Comment:
found_source is enough :smile:

Reply:
I completely agree. Was already like this and I left it unchanged. However, these lines will be removed in a later PR I am preparing.